### PR TITLE
Update Package installer to work for file extensions

### PR DESCRIPTION
**Pull request by SimonDarksideJ**
# XRTK - Mixed Reality Toolkit Change Request

## Overview

Resolves issue with package installer which only currently works with .Asset files, extended to now also work with any file extension required by a package.

Also resolves an issue whereby if the destination folder did not exist, it creates it

## Target of the change:

Is this enhancement for:

- Core (core framework, interfaces and definitions)
- Platform (new or updated platform)

## Related changes

 - https://github.com/XRTK/SDK/pull/174

### CODE REVIEW

**Comment by SimonDarksideJ** [comment]
Double checking but something is missing there

File: XRTK-Core/Packages/com.xrtk.core/Editor/PackageInstaller.cs [hunk 1]
```diff
@@ -29,11 +29,11 @@ public static class PackageInstaller
         /// <param name="destinationPath">The destination path, typically inside the projects "Assets" directory.</param>
         /// <param name="regenerateGuids">Should the guids for the copied assets be regenerated?</param>
         /// <returns>Returns true if the profiles were successfully copies, installed, and added to the <see cref="MixedRealityToolkitRootProfile"/>.</returns>
-        public static bool TryInstallAssets(string sourcePath, string destinationPath, bool regenerateGuids = true)
+        public static bool TryInstallAssets(string sourcePath, string destinationPath, string extension = "asset", bool regenerateGuids = true)
```
Comment (StephenHodgson) [inline XRTK-Core/Packages/com.xrtk.core/Editor/PackageInstaller.cs:32]: I'd rather just use the search options like in the guid regeneration script and not add the additional parameter.

File: XRTK-Core/Packages/com.xrtk.core/Editor/PackageInstaller.cs [hunk 2]
```diff
@@ -43,9 +43,13 @@ public static bool TryInstallAssets(string sourcePath, string destinationPath, b
                 EditorApplication.delayCall += () => AddConfigurations(installedAssets);
                 return true;
             }
+            else
+            {
+                Directory.CreateDirectory(Path.GetFullPath(destinationPath));
```
Comment (StephenHodgson) [inline XRTK-Core/Packages/com.xrtk.core/Editor/PackageInstaller.cs:48]: I think we should always do this at the top of the method if the directory doesn't exist

**Comment by StephenHodgson** [comment]
Link to ask PR is broken

File: XRTK-Core/Packages/com.xrtk.core/Editor/PackageInstaller.cs (hunk 2 above)
Comment (SimonDarksideJ) [reply XRTK-Core/Packages/com.xrtk.core/Editor/PackageInstaller.cs:48]: It does?  The first line checks if the directory exists.  This is then the "Else" on that check?

File: XRTK-Core/Packages/com.xrtk.core/Editor/PackageInstaller.cs (hunk 1 above)
Comment (SimonDarksideJ) [reply XRTK-Core/Packages/com.xrtk.core/Editor/PackageInstaller.cs]: Updated all file functions to use the same method for gathering all known Unity files

File: XRTK-Core/Packages/com.xrtk.core/Editor/Utilities/UnityFileHelper.cs [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+            "*.anim",
+            "*.prefab",
+            "*.unity",
+            "*.asset",
+            "*.guiskin",
+            "*.fontsettings",
+            "*.controller",
+        };
+
+        public static List<string> GetUnityFiles(string assetsRootPath)
```
Comment (StephenHodgson) [inline XRTK-Core/Packages/com.xrtk.core/Editor/Utilities/UnityFileHelper.cs:23]: Should we also make sure this utility can handle relative/absolute paths?

Comment (StephenHodgson) [reply XRTK-Core/Packages/com.xrtk.core/Editor/Utilities/UnityFileHelper.cs:23]: Are we expecting full paths back or relative paths back?

Comment (StephenHodgson) [reply XRTK-Core/Packages/com.xrtk.core/Editor/Utilities/UnityFileHelper.cs:23]: Also maybe a better name would be `GetUnityAssetsAtPath` or `GetUnityAssetsInDirectory` and pass in a `DirectoryInfo` instead?

File: XRTK-Core/Packages/com.xrtk.core/Editor/PackageInstaller.cs [hunk 4]
```diff
@@ -119,7 +123,6 @@ private static string CopyAsset(this string rootPath, string sourceAssetPath, st
                 Directory.CreateDirectory(Directory.GetParent(destinationPath).FullName);
 
                 File.Copy(sourceAssetPath, destinationPath);
-                File.Copy($"{sourceAssetPath}{META_SUFFIX}", $"{destinationPath}{META_SUFFIX}");
```
Comment (StephenHodgson) [inline XRTK-Core/Packages/com.xrtk.core/Editor/PackageInstaller.cs:122]: We should def be copying the meta files too.

File: XRTK-Core/Packages/com.xrtk.core/Editor/PackageInstaller.cs (hunk 2 above)
Comment (StephenHodgson) [reply XRTK-Core/Packages/com.xrtk.core/Editor/PackageInstaller.cs:48]: Ah you're right. we have a different path if it exists.

File: XRTK-Core/Packages/com.xrtk.core/Editor/Utilities/UnityFileHelper.cs (hunk 3 above)
Comment (SimonDarksideJ) [reply XRTK-Core/Packages/com.xrtk.core/Editor/Utilities/UnityFileHelper.cs:23]: Everywhere this is used, we only have a string path.  Might consider an overload to use DirectoryInfo, but nowhere would use it at present.  So this is fine.

OK, Will rename to GetUnityAssetsAtPath

File: XRTK-Core/Packages/com.xrtk.core/Editor/PackageInstaller.cs (hunk 4 above)
Comment (SimonDarksideJ) [reply XRTK-Core/Packages/com.xrtk.core/Editor/PackageInstaller.cs:122]: Please read the full scope of the change before commenting.  .meta is one of the file types ALREADY included.  I copied the method used to regenerate GUIDs on your suggestion and that already included Meta files.

Check in the UnityFileHelper that lists all the file types included

**Comment by SimonDarksideJ** [comment]
As stated, the GetUnityAssetsAtPath function now copies the Meta files, so no need to copy them twice!

**Comment by StephenHodgson** [comment]
Nice, I didn't catch that right away. Great work.

Should we also address the guid regeneration after we copied all the files from multiple directories for a package in this PR?

**Comment by SimonDarksideJ** [comment]
Ok, will test and see if we can regen the guids after the folders.

File: XRTK-Core/Packages/com.xrtk.core/Editor/Utilities/UnityFileHelper.cs [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+            "*.anim",
+            "*.prefab",
+            "*.unity",
+            "*.asset",
+            "*.guiskin",
+            "*.fontsettings",
+            "*.controller",
+        };
+
+        public static List<string> GetUnityAssetsAtPath(string assetsRootPath)
```
Comment (StephenHodgson) [inline XRTK-Core/Packages/com.xrtk.core/Editor/Utilities/UnityFileHelper.cs:23]: Can we add some documentation here?

I'd like to explicitly say that we're returning full paths

Comment (SimonDarksideJ) [reply XRTK-Core/Packages/com.xrtk.core/Editor/Utilities/UnityFileHelper.cs:31]: Done

File: XRTK-Core/Packages/com.xrtk.core/Editor/Utilities/UnityFileHelper.cs [hunk 6]
```diff
@@ -8,6 +8,9 @@ namespace XRTK.Editor.Utilities
 {
     public static class UnityFileHelper
     {
+        /// <summary>
+        /// Managed list of recognised Unity files
```
Comment (StephenHodgson) [inline XRTK-Core/Packages/com.xrtk.core/Editor/Utilities/UnityFileHelper.cs:12]:
```suggestion
        /// Managed list of native Unity asset file extensions
```

**Comment by StephenHodgson** [comment]
Now that I think about it, I think if we fail at any point I'd rather just delete any assets we've copied so far.

**Comment by StephenHodgson** [comment]
Don't forget to add the link to the sdk pull request as well please

File: XRTK-Core/Packages/com.xrtk.core/Editor/PackageInstaller.cs [hunk 7]
```diff
@@ -186,5 +187,10 @@ public static void InstallConfiguration(MixedRealityPlatformServiceConfiguration
             AssetDatabase.SaveAssets();
             EditorApplication.delayCall += () => AssetDatabase.Refresh(ImportAssetOptions.ForceUpdate);
         }
+
+        public static void RegenerateGuids(string[] filePaths)
+        {
+            GuidRegenerator.RegenerateGuids(filePaths, false);
```
Comment (StephenHodgson) [inline XRTK-Core/Packages/com.xrtk.core/Editor/PackageInstaller.cs:193]: Would this make more sense to be an overload on `GuidRegenerator` itself?

Comment (StephenHodgson) [reply XRTK-Core/Packages/com.xrtk.core/Editor/PackageInstaller.cs:193]: AH I see, I made it internal so people wouldn't use it outside of the core project.

**Comment by SimonDarksideJ** [comment]
Will retest in the morning to see if the new updates don't break the fixes

**Pull request merged**